crates/bevy_solari/src/realtime/restir_gi.wgsl

```diff
@@ -295,16 +276,124 @@ fn merge_reservoirs(
 
         let inverse_target_function = select(0.0, 1.0 / other_target_function, other_target_function > 0.0);
         combined_reservoir.unbiased_contribution_weight = combined_reservoir.weight_sum * inverse_target_function;
-
-        return ReservoirMergeResult(combined_reservoir, other_radiance);
     } else {
         combined_reservoir.sample_point_world_position = canonical_reservoir.sample_point_world_position;
         combined_reservoir.sample_point_world_normal = canonical_reservoir.sample_point_world_normal;
         combined_reservoir.radiance = canonical_reservoir.radiance;
 
         let inverse_target_function = select(0.0, 1.0 / canonical_target_function, canonical_target_function > 0.0);
         combined_reservoir.unbiased_contribution_weight = combined_reservoir.weight_sum * inverse_target_function;
+    }
+
+    return combined_reservoir;
+}
+
+struct ReservoirMergeResult {
+    merged_reservoir: Reservoir,
+    selected_sample_radiance: vec3<f32>,
+}
+
+fn merge_reservoirs_spatial(
+    canonical_reservoir: Reservoir,
+    canonical_world_position: vec3<f32>,
+    canonical_world_normal: vec3<f32>,
+    canonical_diffuse_brdf: vec3<f32>,
+    other_reservoir: Reservoir,
+    other_world_position: vec3<f32>,
+    other_world_normal: vec3<f32>,
+    other_diffuse_brdf: vec3<f32>,
+    rng: ptr<function, u32>,
+) -> ReservoirMergeResult {
+    // Radiances for resampling
+    let canonical_sample_radiance =
+        canonical_reservoir.radiance *
+        saturate(dot(normalize(canonical_reservoir.sample_point_world_position - canonical_world_position), canonical_world_normal)) *
+        canonical_diffuse_brdf;
+    let other_sample_radiance =
+        other_reservoir.radiance *
+        saturate(dot(normalize(other_reservoir.sample_point_world_position - canonical_world_position), canonical_world_normal)) *
+        canonical_diffuse_brdf *
+        trace_point_visibility(canonical_world_position, other_reservoir.sample_point_world_position);
+
+    // Target functions for resampling and MIS
+    let canonical_target_function_canonical_sample = luminance(canonical_sample_radiance);
+    let canonical_target_function_other_sample = luminance(other_sample_radiance);
+
+    // Extra target functions for MIS
+    let other_target_function_canonical_sample = luminance(
+        canonical_reservoir.radiance *
+        saturate(dot(normalize(canonical_reservoir.sample_point_world_position - other_world_position), other_world_normal)) *
+        other_diffuse_brdf
+    );
+    let other_target_function_other_sample = luminance(
+        other_reservoir.radiance *
+        saturate(dot(normalize(other_reservoir.sample_point_world_position - other_world_position), other_world_normal)) *
+        other_diffuse_brdf
```

Comment:
this duplication can be extracted into a function to dedup

Reply:
Nah I'd rather keep it explicit here. There's a lot of different things you can do (trace extra rays for visibility, lots of different pairwise MIS schemes) and stuff and it gets complicated. I don't want to make it harder to change in the future.